crates/sui-types/src/certificate_proof.rs

```diff
@@ -1,33 +1,38 @@
 // Copyright (c) Mysten Labs, Inc.
 // SPDX-License-Identifier: Apache-2.0
 
-use crate::committee::EpochId;
 use crate::messages_checkpoint::CheckpointSequenceNumber;
+use crate::{committee::EpochId, crypto::AuthorityStrongQuorumSignInfo};
 
 use serde::{Deserialize, Serialize};
 
-/// CertificateProof is a placeholder for signatures, which indicates that the wrapped message has
-/// been proven valid through indirect means, typically inclusion in a certified checkpoint or
-/// via f+1 votes that the message is correct.
-#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
-pub struct CertificateProof(CertificateProofKind);
+/// CertificateProof is a proof that a transaction certs existed at a given epoch and hence can be executed.
+/// There are two types of proofs: one that is proven by inclusion in a checkpoint and one that is proven by quorum signature.
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub enum CertificateProof {
+    /// Validity was proven by inclusion in the given checkpoint
+    Checkpoint(EpochId, CheckpointSequenceNumber),
+    /// Validity was proven by transaction certificate signature
+    Certified(AuthorityStrongQuorumSignInfo),
+    /// At least f+1 validators have executed this transaction.
+    /// In practice, we will always get 2f+1 (effects cert), but theoretically f+1 is enough to prove
+    /// that the transaction is valid.
+    QuorumExecuted(EpochId),
```

Comment:
I don't think we are likely to need QuorumExecuted any more, now that validators do not do causal completion. Can we remove it?

Reply:
We actually need it, for the case where a fullnode tries to execute a transaction locally. Such transaction may not come with a certificate, but only effects cert.